src/journal/Future.h

```diff
@@ -18,13 +19,16 @@ class FutureImpl;
 
 class Future {
 public:
-  typedef boost::intrusive_ptr<FutureImpl> FutureImplPtr;
-
-  Future() {}
-  Future(const FutureImplPtr &future_impl) : m_future_impl(future_impl) {}
-
-  inline bool is_valid() const {
-    return m_future_impl.get() != nullptr;
+  Future();
```

Comment:
why do we need to explicitly define these constructor? and assignment operator?

as i think compiler will just do this for us as `boost::intrusive_ptr<FutureImpl>` already implements all that are needed.

Reply:
The default constructor doesn't know the size of `m_future_impl` to free it. So we need to define them in the `.cc` file.

Reply:
@batrick if `m_future_impl` is an incomplete type in .h. probably we can just use `= default` for defining those constructor and assignment operator instead if handcrafting them?

Reply:
You mean `= default` in the .cc file?

Reply:
@batrick exactly.

Reply:
Ah, I didn't know that could be valid. I'll give it a try.